carbon-design-system/carbon
# feat(text-input): add readonly variant

Ref #2177, #3018, #3054

Adds a readonly variant to the TextInput component.

Intentional differences from the original design spec:
- No tooltip on the `EditOff16` icon
  - Looking at the planned changes of the tooltip component in v11, adding full support for that at this time seems inefficient. It would require the addition of multiple props that aren't really in the scope of the text input but rather of the tooltip (`tooltipAlignment`, `tooltipPosition`) and might introduce accessiblity issues that are to be addressed with the tooltip changes in v11.
  - In the meantime, teams can still inform their users about the reason why a field is read-only with a couple of options:
    - Use the helper text as explaination
    - Include a tooltip in the form label (quick poc with disabled field: https://jfckr.csb.app/)
- No tooltip on the entire input to show overflow text
  - Same as above
  - Users have the ability to "scroll" inside the field with both their mouse or their keyboard (arrow keys) to reveal overflowing text (native browser behaviour)

#### Changelog

**New**

- `props.readonly` on `TextInput`

#### Testing / Reviewing

- Ensure `readonly` attribute is set to `<input>` accordingly
- Ensure no visual regressions appear with any combination of `invalid`, `warn`, `readonly` on both the `TextInput` and the `PasswordInput` component
- Ensure that no error states are shown when a field is marked as readonly

Side note: this PR also adds a helper `useNormalizedInputProps` which aims to reduce some of the many conditional renderings that were introduced over time (e.g. `{!invalid && !warn && !isFluid && !inline && helper}`) by outsourcing them and providing a single entry for the icons and validation messages.
The intent is to add readonly support to other inputs as well (like `NumberInput`). This utility should help there as well and streamline the inputs. Let me know what you think!
